vllm_spyre/multimodal/mm_mappings/llava_next.py

```diff
@@ -0,0 +1,158 @@
+        kv_cache_specs['head_dim'] = getattr(
+            self.fms_config.text_config, "head_dim",
+            self.hf_config.text_config.hidden_size //
+            self.hf_config.text_config.num_attention_heads)
+        return kv_cache_specs
+
+    @staticmethod
+    def get_mm_specific_load_overrides(hf_config: PretrainedConfig):
+        """Get any overrides needed for initializing the FMS model from the
+        transformers config. For this model, we need to fix the head_dim, which
+        currently surfaces as a problem for all 2b variants of granite 3.x LLMs
+        when running through FMS.
+        
+        TODO: If additional variants of granite vision are added, or broader
+        llava next support is added in FMS, handle it properly here.
+        """
+        serialization.extend_adapter(
+            "llava_next", "hf", ["weight_expansion_for_mismatched_head_dim"])
+        return {
+            "override_hf_pretrained_config": True,
+            "text_config": {
+                "head_dim": 128
+            },
+        }
+
+    @staticmethod
+    def get_maybe_mm_embeddings(
+        fms_model: torch.nn.Module,
+        input_ids: torch.Tensor,
+        mm_features: list[MultiModalFeatureSpec],
+        is_decode: bool,
+    ) -> torch.Tensor:
+        """Get the text or multimodal embeddings for Llava Next using
+        the (potentially compiled) FMS model.
+        """
+        fms_kwargs = {"use_cache": True}
+        mm_spec_keys = ["pixel_values", "image_sizes"]
+
+        # Only merge multimodal features in prefill; nothing mm in decode
+        if mm_features:
+            if len(mm_features) != 1:
+                raise ValueError(
+                    "Currently we assume we only embed one mm request at a time"
+                )
+            mm_spec = mm_features[0].data
+            if mm_spec is not None:
+                # NOTE: This should be pretty safe as it's dependent on the
+                # vLLM/HF processor objects, but we check it anyway to be safe
+                # for now, since transformers 5.0 is just around the corner.
+                if any(k not in mm_spec for k in mm_spec_keys):
+                    raise KeyError(
+                        f"Llava Next requires kwargs: {mm_spec_keys}")
+
+                fms_kwargs["pixel_values"] = mm_spec["pixel_values"].data
+                image_sizes = mm_spec["image_sizes"].data
+
+                # Careful about this; if it's 1D, we'll a tensor of shape
+                # [x, y], which will break in a weird way in image packing,
+                # since it assumes it's 2D and will get sad about getting
+                # an int instead of an iterable
+                if image_sizes.ndim == 1:
+                    image_sizes = image_sizes.unsqueeze(0)
+                fms_kwargs["image_sizes"] = image_sizes
+
+        # NOTE: use_cache is actually not used here, but currently it's
```

Comment:
What does `not used here` mean?

I see a code path that uses it: https://github.com/foundation-model-stack/foundation-model-stack/blob/v1.5.0/fms/models/llava_next.py#L392-L394

Or do you mean that there's no caching actually happening?